### PR TITLE
Do not attempt to move diagnostics with additional locations.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/10258

The issue here is that the "InvokeDelegateWithConditionalAccess" analyzer uses 'additional locations' on diagnostic datas to pass along data.  Unfortunately, the diagnostic analyzer infrastructure was not properly updating these spans when it attempted to preserve and move around diagnostics.

Updating these diagnostics is tricky to do properly.  Additional locations can span across documents or may be in arbitrary locations in the same docoument as the diagnostic.  As such, if have those diagnostics, we don't try to reuse and move the owning diagnostics.  Instead, we reacquire these diagnostics so that all the data in them is correct.

Because of the complexity of this repro the testing will be done with an integration test in roslyn-internal.

### CODE REVIEW

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs [hunk 1]
```diff
@@ -111,6 +125,21 @@ public AnalyzerExecutor(DiagnosticIncrementalAnalyzer owner)
...
 
+            private bool CanReuseDiagnostic(DiagnosticData diagnostic)
+            {
+                // We can't reuse diagnostics that contain additional locations.  It's too 
+                // difficult to ensure that all the locations will be moved to the right
+
+                return diagnostic.AdditionalLocations == null || diagnostic.AdditionalLocations.Count == 0;
+            }
+
+            private static async Task<ImmutableArray<DiagnosticData>> UpdateAllSemanticDiagnostics(StateSet stateSet, DiagnosticAnalyzerDriver analyzerDriver, ImmutableArray<DiagnosticData> diagnosticData)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs]: Suffix with Async?

**Comment by mavasani** [comment]
Tagging @heejaechang so the v2 OOP engine also handles this.

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs]: Sure.

**Comment by CyrusNajmabadi** [comment]
Ping @dotnet/roslyn-ide

**Comment by jmarolf** [comment]
Can you point us at the PR that contains the internal tests?

**Comment by CyrusNajmabadi** [comment]
@jmarolf it's here: https://github.com/dotnet/roslyn-internal/pull/788

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by jmarolf** [comment]
:+1:

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs [hunk 2]
```diff
@@ -94,15 +95,28 @@ public AnalyzerExecutor(DiagnosticIncrementalAnalyzer owner)
                     {
                         var memberDxData = await GetSemanticDiagnosticsAsync(analyzerDriver, stateSet.Analyzer).ConfigureAwait(false);
 
-                        diagnosticData = _owner.UpdateDocumentDiagnostics(existingData, ranges.Ranges, memberDxData.AsImmutableOrEmpty(), root.SyntaxTree, member, memberId);
-                        ValidateMemberDiagnostics(stateSet.Analyzer, document, root, diagnosticData);
+                        // if all the current diagnostics have all their locations in this document,
```
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs]: I think this method shouldn't be called for analyzer that can't do member update. just make the analyzer never to be in this method.

Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.AnalyzerExecutor.cs]: hmmm.. this code seems changed quite a bit over what I remember. anyway, v2 engine doesn't make span based analysis anymore so it shouldn't be an issue for v2.

**Comment by heejaechang** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Pull request closed without merging**